quinn/examples/common.rs

```diff
@@ -0,0 +1,79 @@
+//! Commonly used code in most examples.
+
+use failure::Error;
+use quinn::{
+    Certificate, CertificateChain, ClientConfig, ClientConfigBuilder, Driver, Endpoint, Incoming,
+    PrivateKey, ServerConfig, ServerConfigBuilder, TransportConfig,
+};
+use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
+use std::sync::Arc;
+
+/// Returns QUIC endpoint builder configured to build a client endpoint.
+///
+/// ## Args
+///
+/// - server_certs: list of trusted certificates.
+#[allow(unused)]
+pub fn make_client_endpoint(
+    bind_addr: &SocketAddr,
+    server_certs: &Vec<Vec<u8>>,
+) -> Result<(Endpoint, Driver), Error> {
+    let client_cfg = configure_client(server_certs)?;
+    let mut endpoint_builder = Endpoint::new();
+    endpoint_builder.default_client_config(client_cfg);
+    let (endpoint, driver, _incoming) = endpoint_builder.bind(bind_addr)?;
+    Ok((endpoint, driver))
+}
+
+/// Configures server to listen on given address and port.
+///
+/// ## Returns
+/// - UDP socket driver
+/// - a sream of incoming QUIC connections
+/// - server certificate serialized into DER format
+pub fn make_server_endpoint(bind_addr: &SocketAddr) -> Result<(Driver, Incoming, Vec<u8>), Error> {
+    let (server_config, server_cert) = configure_server()?;
+    let mut endpoint_builder = Endpoint::new();
+    endpoint_builder.listen(server_config);
+    let (_endpoint, driver, incoming) = endpoint_builder.bind(bind_addr)?;
+    Ok((driver, incoming, server_cert))
+}
+
+/// Convenience function to build IPv4 address with port.
+pub fn ipv4_addr(a: u8, b: u8, c: u8, d: u8, port: u16) -> SocketAddr {
+    SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::new(a, b, c, d), port))
+}
```

Comment:
Would it make sense to make this unnecessary by modifying functions to take an impl of [ToSocketAddrs](https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html) instead of a concrete `SocketAddr`? That's what `bind` takes, allowing traditional socket address syntax to be supplied as strings.

Reply:
I would prefer to use [net-literals](https://crates.io/crates/net-literals) crate. Just didn't want to introduce any new dependencies without confirmation. Would you be okay with it as a dev dependency?

Reply:
Actually, let me see what I can do with the `ToSocketAddrs` trait - it seems pretty neat, thx for bringing this up :)

Reply:
Yeah, I would say using net-literals for this is somewhat idiosyncratic, which is not a great property for example code.